core/src/apps/cardano/certificates.py

```diff
@@ -0,0 +1,216 @@
+        pool_parameters = certificate.pool_parameters
+
+        return (
+            certificate.type,
+            pool_parameters.pool,
+            pool_parameters.vrf_key_hash,
+            pool_parameters.pledge,
+            pool_parameters.cost,
+            cbor.Tagged(
+                30,
+                (
+                    pool_parameters.margin_numerator,
+                    pool_parameters.margin_denominator,
+                ),
+            ),
+            pack_reward_address_bytes(
+                pool_parameters.reward_account_key_hash, network_id
+            ),
+            _cborize_pool_owners(keychain, pool_parameters.owners),
+            _cborize_pool_relays(pool_parameters.relays),
+            _cborize_pool_metadata(pool_parameters.metadata),
+        )
+    else:
+        raise INVALID_CERTIFICATE
+
+
+def _validate_pool_parameters(pool_parameters: CardanoPoolParametersType):
+    if any(
+        [
+            not pool_parameters.pool or len(pool_parameters.pool) != POOL_HASH_SIZE,
+            (
+                not pool_parameters.vrf_key_hash
+                or len(pool_parameters.vrf_key_hash) != VRF_KEY_HASH_SIZE
+            ),
+            not 0 <= pool_parameters.pledge < LOVELACE_MAX_SUPPLY,
+            not 0 <= pool_parameters.cost < LOVELACE_MAX_SUPPLY,
+            pool_parameters.margin_numerator <= 0,
+            pool_parameters.margin_denominator <= 0,
+            pool_parameters.margin_numerator > pool_parameters.margin_denominator,
+            (not pool_parameters.owners or len(pool_parameters.owners) == 0),
+        ]
+    ):
+        raise INVALID_CERTIFICATE
+
+    for pool_relay in pool_parameters.relays:
+        _validate_pool_relay(pool_relay)
+
+    _validate_pool_owners(pool_parameters.owners)
+
+    if pool_parameters.metadata:
+        _validate_pool_metadata(pool_parameters.metadata)
+
+
+def _validate_pool_owners(owners: List[CardanoPoolOwnerType]):
+    owners_as_path_count = 0
+    for owner in owners:
+        if any(
+            [
+                not owner.staking_key_hash and not owner.staking_key_path,
+                (
+                    owner.staking_key_hash
+                    and len(owner.staking_key_hash) != PUBLIC_KEY_HASH_SIZE
+                ),
+                (
+                    owner.staking_key_path
+                    and not is_staking_path(owner.staking_key_path)
+                ),
+            ]
+        ):
+            raise INVALID_CERTIFICATE
+
+        if owner.staking_key_path:
+            owners_as_path_count += 1
+
+    if owners_as_path_count != 1:
+        raise INVALID_CERTIFICATE
+
+
+def _validate_pool_relay(pool_relay: CardanoPoolRelayParametersType):
+    if pool_relay.type not in [
+        CardanoPoolRelayType.SINGLE_HOST_IP,
+        CardanoPoolRelayType.SINGLE_HOST_NAME,
+        CardanoPoolRelayType.MULTIPLE_HOST_NAME,
+    ]:
+        raise INVALID_CERTIFICATE
+
```

Comment:
`dns_name` is required for `SINGLE_HOST_NAME` and `MULTIPLE_HOST_NAME`, so we should check it is present and has proper length?

Reply:
This is not something the user would be validating anyway as it's operator's concern, but I added the validation here for the sake of completeness